src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -881,4 +883,152 @@ protected static async Task<ImmutableArray<string>> GetAllMatchingGlobalAliasNam
     }
+
+    /// <summary>
+    /// Discover implied symbols, which are defined based on the requested symbol.
+    /// Implied symbols include anonymous type properties and tuple fields.
+    /// </summary>
+    protected static async ValueTask DiscoverImpliedSymbolsAsync(
+        TSymbol symbol, Solution solution, ArrayBuilder<ISymbol> symbolBuilder, CancellationToken cancellationToken)
+    {
+        using var _ = PooledHashSet<ISymbol>.GetInstance(out var symbolSet);
+
+        await ProducerConsumer<Document>.RunParallelAsync(
+            GetCandidateDocumentsAsync(),
+            produceItems: static async (document, callback, args, cancellationToken) =>
+            {
+                var (symbol, symbolSet) = args;
+                var name = symbol.Name;
+
+                var shouldProcess = await ShouldProcessDocumentAsync(document, name, cancellationToken).ConfigureAwait(false);
+                if (shouldProcess)
+                {
+                    callback(document);
+                }
+            },
+            consumeItems: static async (documents, args, cancellationToken) =>
+            {
+                var (symbol, symbolSet) = args;
+                await foreach (var document in documents)
+                {
+                    await FindImplicitSymbolsInDocumentAsync(symbol, symbolSet, document, cancellationToken).ConfigureAwait(false);
+                }
+            },
+            args: (symbol, symbolSet),
+            cancellationToken).ConfigureAwait(false);
+
+        symbolBuilder.AddRange(symbolSet);
+
+        async IAsyncEnumerable<Document> GetCandidateDocumentsAsync()
+        {
+            var name = symbol.Name;
+
+            foreach (var project in solution.Projects)
+            {
+                var allDocuments = project.GetAllRegularAndSourceGeneratedDocumentsAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                await foreach (var document in allDocuments)
+                {
+                    yield return document;
+                }
+            }
+        }
+
+        static async ValueTask<bool> ShouldProcessDocumentAsync(Document document, string name, CancellationToken cancellationToken)
+        {
+            if (!document.SupportsSemanticModel)
+                return false;
+
+            var treeIndex = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (treeIndex is null)
+                return false;
+
+            return treeIndex.ProbablyContainsIdentifier(name);
+        }
+    }
+
+    private static async ValueTask FindImplicitSymbolsInDocumentAsync(TSymbol symbol, PooledHashSet<ISymbol> symbolSet, Document document, CancellationToken cancellationToken)
+    {
+        var semanticModel = await document!.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
This breaks find references. A core design of it is that it only ever once creates a semantic model for a document, and it does *all* the work it needs on that document at that point.